File: code/ch18-e2e/backend/src/licenseService.js
// src/licenseService.js -- License operations (from book Chapter 18)
//
// Maps REST operations to JSON Ledger API calls.
// Matches the code listing in the book exactly.

import { randomUUID } from "node:crypto";
import {
  submitCommand,
  getActiveContracts,
} from "./ledgerClient.js";

const TEMPLATE =
  "Main.License:License";

export async function createLicense(
  token, vendor, customer,
  product, expiresAt
) {
  return submitCommand(
    token,
    [vendor],
    [{
      CreateCommand: {
        templateId: TEMPLATE,
        createArguments: {
          vendor, customer, product,
          issuedAt: new Date().toISOString(),
          expiresAt,
        },
      },
    }],
    `create-license-${randomUUID()}`
  );
}

export async function renewLicense(
  token, vendor, contractId, newExpiry
) {
  return submitCommand(
    token,
    [vendor],
    [{
      ExerciseCommand: {
        templateId: TEMPLATE,
        contractId,
        choice: "Renew",
        choiceArgument: {
          newExpiry,
        },
      },
    }],
    `renew-license-${randomUUID()}`
  );
}

export async function listLicenses(
  token, party
) {
  return getActiveContracts(
    token, party, TEMPLATE
  );
}

File: book-1-daml-canton-guide/code/ch18-e2e/backend-java/src/main/java/com/example/license/LicenseService.java
// LicenseService.java
package com.example.license;

import com.fasterxml.jackson.databind.*;
import com.fasterxml.jackson.databind.node.*;
import java.util.UUID;
import org.springframework.stereotype.Service;

@Service
public class LicenseService {
    private static final String TEMPLATE =
        "Main.License:License";
    private final LedgerClient ledger;
    private final ObjectMapper mapper =
        new ObjectMapper();

    public LicenseService(LedgerClient ledger) {
        this.ledger = ledger;
    }

    public JsonNode createLicense(
        String token,
        String vendor,
        String customer,
        String product,
        String expiresAt
    ) throws Exception {
        var args = mapper.createObjectNode();
        args.put("vendor", vendor);
        args.put("customer", customer);
        args.put("product", product);
        args.put("issuedAt",
            java.time.Instant.now().toString());
        args.put("expiresAt", expiresAt);

        var cmd = mapper.createObjectNode();
        var create = mapper.createObjectNode();
        create.put("templateId", TEMPLATE);
        create.set("createArguments", args);
        cmd.set("CreateCommand", create);

        return ledger.submitCommand(
            token,
            new String[]{ vendor },
            new ObjectNode[]{ cmd },
            "create-license-" + UUID.randomUUID()
        );
    }

    public JsonNode renewLicense(
        String token,
        String vendor,
        String contractId,
        String newExpiry
    ) throws Exception {
        var cmd = mapper.createObjectNode();
        var exercise = mapper.createObjectNode();
        exercise.put("templateId", TEMPLATE);
        exercise.put("contractId", contractId);
        exercise.put("choice", "Renew");
        var choiceArg = mapper.createObjectNode();
        choiceArg.put("newExpiry", newExpiry);
        exercise.set("choiceArgument", choiceArg);
        cmd.set("ExerciseCommand", exercise);

        return ledger.submitCommand(
            token,
            new String[]{ vendor },
            new ObjectNode[]{ cmd },
            "renew-license-" + UUID.randomUUID()
        );
    }

    public JsonNode listLicenses(
        String token, String party
    ) throws Exception {
        return ledger.getActiveContracts(
            token, party, TEMPLATE
        );
    }
}
